### PR TITLE
Rtl type for copter

### PR DESCRIPTION
**Pull request by khancyr**
This PR intend to add RTL type to allow different strategie for RTL.
Currently it implement two new simple case : always go home, always go to rally.

Coming soon : RTL doing automission in backward. and same thing for Rover !

### CODE REVIEW

**Comment by Pedals2Paddles** [comment]
These are all things I've repeatedly had use cases for and wished I could do!  Always going to a Rally Point is very useful. Many times I've had a takeoff location that I have to navigate hand flying due to obstructions.  I can manually takeoff and manually land by flying around the obstacles easily.  But an automated RTL simply dropping down vertically would crash. And not reliable enough for SRTL.  Setting a nearby clear but inconvenient location as a rally point solves that problem regardless of the copter's proximity.  If you've selected Smart RTL as the failsafe, I presume that would bypass any rally point regardless of this setting?

Flying a mission backwards has some unique challenges to overcome I can think of.
- Does it fly back at the RTL speed or at all the various mission speeds. The RTL Speed may not be appropriate for the mission course.  But the mission speeds might be detrimentally slow.
- Does it bypass all the mission commands that are not a course of flight waypoint? I would think so.
- What if you started a mission in-flight rather that from a takeoff?  When it reaches WP1, would it then execute the traditional RTL? 
- Does SmartRTL still record during missions (I think it does)?  If so, is handing mission reversal even necessary when SRTL can do it more efficiently all the way back to the point of takeoff regardless of mission origin?

**Comment by khancyr** [comment]
@Pedals2Paddles, nice to get you back ! 
About rally, the behavior is the same as setting RALLY_INCL_HOME to 0.
About the smartrtl only works from the starting location and as the road is build from it, it always return to start point. I may add another logic to start it from chosen point. That you can do mission, exit mission to fly then smartrlt to last mission wp passed and final rewind the mission :- P

About flying mission backward : 
- it get back on AUTO so fly at mission speed.
- You will be able to chose the behavior, even in a failsafe it means nothing to redo all non-nav cmd.
- On failsafe, yes. On just mission reverse, it will loiter at first point.
- On long mission I am not sure Smartrtl is better as it is limited in point. Maybe the optimisation loop on the mission file could be done too to get a simplier path.

**Comment by rmackay9** [comment]
It's always possible to make the vehicle fly to a rally point (and not home) by setting the [RALLY_INC_HOME](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Rally/AP_Rally.cpp#L45) parameter to 0.

I'm a little skeptical of the usefulness of this feature.  Also is it RTL that we want to be enhanced or is it that we want a failsafe feature that executes a mission?  If it's the latter then I think it's already handled as part of the advanced failsafe library.

If the issue is that SmartRTL can't store enough points, then let's address that issue directly instead of developing a similar feature.

**Comment by rmackay9** [comment]
After a bit more thought, I take back some of what I said above.  I think it could be quite useful for the user to be able to lay out a mission and then when they switch the vehicle into RTL, it finds the closest point on the path, flies directly to it and then follows the mission (in reverse or forwards) to return home.

**Comment by khancyr** [comment]
The sole purpose of this PR is to allow to add different strategie for RTL, the default one is good for 75% of use case until we have some active avoidance and pathplanning. 
I can remove the part on force go on rally, I think too that it is the same as setting RALLY_INC_HOME.

I don't want to discuss here about smartrtl, that is not the point. It is just that smartrtl lack of tests and mostly on long flight time. I think that having a mission wrote and smartrtl a both viable solutions. But let's not discuss that here. Here is just for add possibilitty to add new RTL behaviors

**Comment by Kiwa21** [comment]
Well, the end goal of this PR of @khancyr  (and the following ones), is to have different RTL strategies and be able to choose one regarding the scenario without tweeking parameters in flight.
I think RALLY_INC_HOME is a very bad parameter. HOME and RALLY should always be distincts.
You want the ability to define rally points in case of emergency but you don't want the copter to always come back to a rally point if it can / if the user want it to go home.

Moreover, we will be doing command like a "RTL_GO_TO" with a parameter (type = home, closest rallypoint, specific id rally point).
We will happily discuss this before doing implementation if you want @rmackay9

**Comment by jmachuca77** [comment]
Related to #807

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -988,6 +988,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_SUBGROUPINFO(follow, "FOLL", 27, ParametersG2, AP_Follow),
 #endif
 
+    // @Param: RTL_TYPE
+    // @DisplayName: RTL handling type
+    // @Description: This controls how RTL is handled. A value of zero means a conventional RTL. A value of 1 uses the closest DO_LAND_START marker in the mission waypoints to fly a pre-planned path. If no DO_LAND_START markers a found a conventional RTL is used.
```
Comment (khancyr) [inline ArduCopter/Parameters.cpp]: wrong comments

**Comment by khancyr** [comment]
Closing as I get a much better solution

**Pull request closed without merging**